internal/server/run.go

```diff
@@ -0,0 +1,93 @@
+package server
+
+import (
+	"errors"
+	"io"
+	"log"
+	"net"
+	"tcp-load-balancer/internal/upstream"
+)
+
+var ErrUninitialized = errors.New("load balancer not initialized")
+
+// Run handles incoming connections until terminated.
+func (l *LoadBalancer) Run() error {
+	if l == nil || l.listener == nil {
+		return ErrUninitialized
+	}
+
+	for {
+		// TODO: Set up mTLS in next PR. For now, connect without TLS.
+		clientConn, err := l.listener.Accept()
+		if err != nil {
+			go respondAndClose(clientConn, err.Error())
+			continue
+		}
+
+		if err := l.handleConnection(clientConn); err != nil {
+			log.Printf("Unable to handle connection: %s", err)
+			go respondAndClose(clientConn, err.Error())
+		}
+	}
+}
+
+// handleConnection selects an upstream host, tracks connection counts, and forwards data upstream.
+func (l *LoadBalancer) handleConnection(clientConn net.Conn) error {
+	// Host selection is not included in goroutine handling so that requests arriving at the same time are not routed to the same host.
+	// This adds a small amount of latency to the request, but ensures accurate load balancing.
+	host, err := l.LeastConnections()
+	if err != nil {
+		return err
```

Comment:
Missing clientConn.Close call in case of `l.LeastConnections()` error.

Reply:
The Run func is now handles closing the connection since that's where it is opened.

```
// Run handles incoming connections until terminated.
func (l *LoadBalancer) Run() error {
	if l == nil || l.listener == nil {
		return ErrUninitialized
	}

	for {
		// TODO: Set up mTLS in next PR. For now, connect without TLS.
		clientConn, err := l.listener.Accept()
		if err != nil {
			break
		}

		if err := l.handleConnection(clientConn); err != nil {
			log.Printf("Unable to handle connection: %s", err)
		}
		
		if err := clientConn.Close(); err != nil {
			log.Printf("Unable to close client connection: %s", err)
		}	
	}

	return nil
}
```